// shell/lib/NullStream.js
// Generated by CoffeeScript 1.4.0
var NullStream, events,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

events = require('events');

module.exports = NullStream = (function(_super) {

  __extends(NullStream, _super);

  function NullStream() {
    return NullStream.__super__.constructor.apply(this, arguments);
  }

  NullStream.prototype.readable = true;

  NullStream.prototype.pause = function() {};

  NullStream.prototype.resume = function() {};

  NullStream.prototype.pipe = function() {};

  NullStream.prototype.writable = true;

  NullStream.prototype.write = function(data) {
    return this.emit('data', data);
  };

  NullStream.prototype.end = function() {
    return this.emit('close');
  };

  NullStream.prototype.destroy = function() {};

  NullStream.prototype.destroySoon = function() {};

  return NullStream;

})(events.EventEmitter);
