// each/lib/index.js
// Generated by CoffeeScript 1.10.0
var Each, glob, path, stream, util;

stream = require('stream');

path = require('path');

glob = require('glob');

util = require('util');


/*
each(elements)
.parallel(false|true|integer)
.sync(false)
.times(1)
.repeat(1)
.files(cwd, ['./*.coffee'])
.push(element)
.unshift(element)
.write(element)
.pause()
.resume()
.close()
.call(callback)
.error(callback)
.then(callback)
Chained and parallel async iterator in one elegant function
 */

Each = function(options, _elements) {
  var isObject, self, type;
  this.options = options;
  this._elements = _elements;
  if (arguments.length === 1) {
    this._elements = this.options;
    this.options = {};
  }
  type = typeof this._elements;
  if (this._elements === null || type === 'undefined') {
    this._elements = [];
  } else if (type === 'number' || type === 'string' || type === 'function' || type === 'boolean') {
    this._elements = [this._elements];
  } else if (!Array.isArray(this._elements)) {
    isObject = true;
  }
  if (isObject) {
    this._keys = Object.keys(this._elements);
  }
  this._errors = [];
  this.options.concurrency = 1;
  this.options.sync = false;
  this.options.times = 1;
  this.options.repeat = false;
  this._close = false;
  this._handler_index = 0;
  this.total = this._keys ? this._keys.length : this._elements.length;
  this.listeners = [];
  this._endable = 1;
  this.started = 0;
  this.done = 0;
  this.paused = 0;
  this.readable = true;
  self = this;
  setImmediate((function(_this) {
    return function() {
      return _this._run();
    };
  })(this));
  return this;
};

Each.prototype._has_next_handler = function() {
  var j, len, listener, occurences, ref;
  occurences = 0;
  ref = this.listeners;
  for (j = 0, len = ref.length; j < len; j++) {
    listener = ref[j];
    if (listener[0] !== 'call') {
      continue;
    }
    if (occurences === this._handler_index) {
      return true;
    }
    occurences++;
  }
  return false;
};

Each.prototype._get_current_handler = function() {
  var j, len, listener, occurences, ref;
  occurences = 0;
  ref = this.listeners;
  for (j = 0, len = ref.length; j < len; j++) {
    listener = ref[j];
    if (listener[0] !== 'call') {
      continue;
    }
    if (occurences === this._handler_index) {
      return listener[1];
    }
    occurences++;
  }
  throw Error('No Found Handler');
};

Each.prototype._call_next_then = function(error, count) {
  var i, j, len, listener, occurences, ref, ref1, ref2;
  occurences = 0;
  ref = this.listeners;
  for (i = j = 0, len = ref.length; j < len; i = ++j) {
    listener = ref[i];
    if (listener[0] === 'call') {
      occurences++;
      continue;
    }
    if (listener[0] === 'error' && error && occurences >= this._handler_index) {
      listener[1].call(null, error);
      if (((ref1 = this.listeners[i + 1]) != null ? ref1[0] : void 0) === 'then') {
        continue;
      } else {
        return;
      }
    }
    if (listener[0] === 'then' && occurences >= this._handler_index) {
      if (((ref2 = this.listeners[i - 1]) != null ? ref2[0] : void 0) === 'error') {
        if (!error) {
          listener[1].call(null, count);
        }
      } else {
        listener[1].call(null, error, count);
      }
      return;
    }
  }
  throw Error('Invalid State: error or then not defined');
};

Each.prototype._run = function() {
  var args, err, error, error1, handler, handlers, i, index, j, l, len, self;
  if (this.paused) {
    return;
  }
  handlers = this._get_current_handler();
  error = null;
  if (this._endable === 1 && (this._close || this.done === this.total * this.options.times * handlers.length || (this._errors.length && this.started === this.done))) {
    this._handler_index++;
    if (this._errors.length || !this._has_next_handler()) {
      this.readable = false;
      if (this._errors.length) {
        if (this.options.concurrency !== 1) {
          if (this._errors.length === 1) {
            error = this._errors[0];
          } else {
            error = new Error("Multiple errors (" + this._errors.length + ")");
            error.errors = this._errors;
          }
        } else {
          error = this._errors[0];
        }
      } else {
        args = [];
      }
      this._call_next_then(error, this.done);
      return;
    }
    handlers = this._get_current_handler();
    this._endable = 1;
    this.started = 0;
    this.done = 0;
    this.paused = 0;
    this.readable = true;
  }
  if (this._errors.length !== 0) {
    return;
  }
  while ((this.options.concurrency === true ? (this.total * this.options.times * handlers.length - this.started) > 0 : Math.min(this.options.concurrency - this.started + this.done, this.total * this.options.times * handlers.length - this.started))) {
    if (this._errors.length !== 0) {
      break;
    }
    if (this._close) {
      break;
    }
    if (this.options.repeat) {
      index = this.started % this._elements.length;
    } else {
      index = Math.floor(this.started / (this.options.times * handlers.length));
    }
    this.started += handlers.length;
    try {
      self = this;
      for (i = j = 0, len = handlers.length; j < len; i = ++j) {
        handler = handlers[i];
        l = handler.length;
        if (this.options.sync) {
          l++;
        }
        switch (l) {
          case 1:
            args = [];
            break;
          case 2:
            if (this._keys) {
              args = [this._elements[this._keys[index]]];
            } else {
              args = [this._elements[index]];
            }
            break;
          case 3:
            if (this._keys) {
              args = [this._keys[index], this._elements[this._keys[index]]];
            } else {
              args = [this._elements[index], index];
            }
            break;
          case 4:
            if (this._keys) {
              args = [this._keys[index], this._elements[this._keys[index]], index];
            } else {
              return self._next(new Error('Invalid arguments in item callback'));
            }
            break;
          default:
            return self._next(new Error('Invalid arguments in item callback'));
        }
        if (!this.options.sync) {
          args.push((function() {
            var count;
            count = 0;
            return function(err) {
              if (err) {
                return self._next(err);
              }
              if (++count !== 1) {
                err = new Error('Multiple call detected');
                if (self.readable) {
                  return self._next(err);
                } else {
                  throw err;
                }
              }
              return self._next();
            };
          })());
        }
        err = handler.apply(null, args);
        if (this.options.sync) {
          self._next(err);
        }
      }
    } catch (error1) {
      err = error1;
      if (self.readable) {
        self._next(err);
      } else {
        throw err;
      }
    }
  }
  return null;
};

Each.prototype._next = function(err) {
  if ((err != null) && err instanceof Error) {
    this._errors.push(err);
  }
  this.done++;
  return this._run();
};

Each.prototype.run = function(callback) {
  console.log('DEPRECATED: use `call` instead of `run`');
  return this.call(callback);
};

Each.prototype.call = function(callback) {
  if (!Array.isArray(callback)) {
    callback = [callback];
  }
  this.listeners.push(['call', callback]);
  return this;
};

Each.prototype.then = function(callback) {
  this.listeners.push(['then', callback]);
  return this;
};

Each.prototype.end = function(callback) {
  this.listeners.push(['end', callback]);
  return this;
};

Each.prototype.error = function(callback) {
  this.listeners.push(['error', callback]);
  return this;
};

Each.prototype.end = function() {
  console.log('Function `end` deprecated, use `close` instead.');
  return this.close();
};

Each.prototype.close = function() {
  this._close = true;
  this._next();
  return this;
};

Each.prototype.sync = function(s) {
  this.options.sync = (s != null) || true;
  return this;
};

Each.prototype.repeat = function(t) {
  this.options.repeat = true;
  this.options.times = t;
  if (this._elements.length === 0) {
    this.write(null);
  }
  return this;
};

Each.prototype.times = function(t) {
  this.options.times = t;
  if (this._elements.length === 0) {
    this.write(null);
  }
  return this;
};

Each.prototype.files = function(base, pattern) {
  var j, len, p;
  if (arguments.length === 1) {
    pattern = base;
    base = null;
  }
  if (Array.isArray(pattern)) {
    for (j = 0, len = pattern.length; j < len; j++) {
      p = pattern[j];
      this.files(p);
    }
    return this;
  }
  this._endable--;
  if (base) {
    pattern = path.resolve(base, pattern);
  }
  glob(pattern, (function(_this) {
    return function(err, files) {
      var file, k, len1;
      for (k = 0, len1 = files.length; k < len1; k++) {
        file = files[k];
        _this._elements.push(file);
      }
      _this.total += files.length;
      return setImmediate(function() {
        _this._endable++;
        return _this._run();
      });
    };
  })(this));
  return this;
};

Each.prototype.write = Each.prototype.push = function(item) {
  var l;
  l = arguments.length;
  if (l === 1) {
    this._elements.push(arguments[0]);
  } else if (l === 2) {
    if (!this._keys) {
      this._keys = [];
    }
    this._keys.push(arguments[0]);
    this._elements[arguments[0]] = arguments[1];
  }
  this.total++;
  return this;
};

Each.prototype.unshift = function(item) {
  var index, l;
  l = arguments.length;
  if (this.options.repeat) {
    index = this.started % this._elements.length;
  } else {
    index = Math.floor(this.started / this.options.times);
  }
  if (l === 1) {
    this._elements.splice(index, 0, arguments[0]);
  } else if (l === 2) {
    if (!this._keys) {
      this._keys = [];
    }
    this._keys.splice(index, 0, arguments[0]);
    this._elements[arguments[0]] = arguments[1];
  }
  this.total++;
  return this;
};

Each.prototype.pause = function() {
  return this.paused++;
};

Each.prototype.resume = function() {
  this.paused--;
  return this._run();
};

Each.prototype.parallel = function(mode) {
  if (typeof mode === 'number') {
    this.options.concurrency = mode;
  } else if (mode) {
    this.options.concurrency = mode;
  } else {
    this.options.concurrency = 1;
  }
  return this;
};

module.exports = function(elements) {
  return new Each(elements);
};

module.exports.Each = Each;
